local-cli/core/findPlugins.js

```diff
@@ -52,26 +60,30 @@ const findPluginInFolder = (folder) => {
   return deps.reduce(
     (acc, pkg) => {
       if (isRNPMPlugin(pkg)) {
-        return acc.concat(pkg);
+        acc.commands.push(pkg);
       }
       if (isReactNativePlugin(pkg)) {
         const pkgJson = readPackage(path.join(folder, 'node_modules', pkg));
-        if (!pkgJson) {
-          return acc;
+        if (pkgJson) {
+          acc.commands.push(findPluginsInReactNativePackage(pkgJson));
+          acc.platforms.push(findPlatformsInPackage(pkgJson));
         }
-        return acc.concat(findPluginsInReactNativePackage(pkgJson));
       }
       return acc;
     },
-    []
+    {commands: [], platforms: []}
   );
 };
 
 /**
  * Find plugins in package.json of the given folder
  * @param {String} folder Path to the folder to get the package.json from
- * @type  {Array}         Array of plugins or an empty array if no package.json found
+ * @type  {Object}        Object of commands and platform plugins
  */
 module.exports = function findPlugins(folders) {
-  return uniq(flatten(folders.map(findPluginInFolder)));
+  const plugins = folders.map(findPluginInFolder);
+  return {
+    commands: uniq(flatten(plugins.map(p => flatten(p.commands)))),
+    platforms: uniq(flatten(plugins.map(p => flatten(p.platforms))))
```

Comment:
I am not sure I am getting that line - the `commands` flattening is to make a flat array of all commands (if plugin defines many, we handle that).

What's the purpose of it for `platforms`? I get that we want to get a list of *all* platforms defined throughout the plugins - but when do we use it later?

Reply:
Not sure if the question here is related to `platforms` and why it's mangled with `commands`, or specifically why we `flatten` the `platforms` collection, so I'll address both.

`platforms` is used by `local-cli/core/index.js` and the `link` and `unlink` command to evaluate `projectConfig()`, `dependencyConfig()`, and `linkConfig()`. I would have created a separate module to extract the platform info, but I didn't want to impact CLI perf. If I created a separate module, I would have to read `package.json` for each potential plugin twice (instead of once as is done here).

In terms of why we flatten the collection of platforms - plugins could potentially define multiple platforms (e.g., uwp and win32, or ios and android), so I flatten the list so we have one array of all platforms used for configuration and in `link`.

Reply:
@grabbou - I removed the inner `flatten`. The problem was that I had switched from `concat` to `push`.